File: programa.c
#include <stdio.h>

int main(int argc,char *argv[])
{
    // nao passou parametro, exibe mensagem e sai fora
    if (argc == 1) {
	printf("Faltou parametro (nome do arquivo de imagem)");
	return 1;
    }

    unsigned char buffer[1];
    FILE *ptr;

    // r = leitura e b = binario
    ptr = fopen(argv[1],"rb");

    // erro na leitura do arquivo, nao existe?
    if (ptr == NULL) {
	printf("Falha ao abrir %s (nao existe?)", argv[1]);
        return 2;
    }

    // faz a leitura do primeiro byte
    fread(buffer,sizeof(buffer),1,ptr);
    fclose(ptr);

    // analisa o primeiro byte do arquivo, conforme ele sabemos o tipo
    switch (buffer[0])
    {
      case 0x89: // 137
        printf("%s", "Formato PNG");
        break;

      case 0xFF: // 255
        printf("%s", "Formato JPEG");
        break;

      case 0x49: // 73
        printf("%s", "Formato TIFF");
        break;

      case 0x42: // 66
        printf("%s", "Formato Windows Bitmap");
        break;

      case 0x47: // 71
        printf("%s", "Formato GIF");
        break;

      default:
        printf("%s", "Formato desconhecido");
    }

  return 0;
}

File: Programa.java
import java.io.*;
 
public class Programa {
    
    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Faltou arquivo no parametro");
            System.exit(0);
        }
 
        try {
            InputStream inputStream = new FileInputStream(args[0]);
            int primeiroByte = inputStream.read();
			inputStream.close();
 		
        	// analisa o primeiro byte do arquivo, conforme ele sabemos o tipo
            switch (primeiroByte)
            {
              case 0x89: // 137
                System.out.println("Formato PNG");
                break;
        
              case 0xFF: // 255
                System.out.println("Formato JPEG");
                break;
        
              case 0x49: // 73
                System.out.println("Formato TIFF");
                break;
        
              case 0x42: // 66
                System.out.println("Formato Windows Bitmap");
                break;
        
              case 0x47: // 71
                System.out.println("Formato GIF");
                break;
        
              default:
                System.out.println("Formato desconhecido");
            }
        } catch (IOException ex) {
            ex.printStackTrace();
        }
     }
}